File: bkasm/src/asmvars.c
#include <stdio.h>
#include <string.h>
#include "bkasm.h"
#include "asmvars.h"
#include "errors.h"
#include "hashtable.h"

HashVar asmvars[MAX_VAR_COUNT];

void asmvars_add(const char key[], int val)
{

    if (key == 0 || strlen(key) == 0)
    {
        fprintf(stderr, "INTERNAL_ERROR: asmvars_add: key is null");
        return;
    }
    hash_push(key, val, asmvars, MAX_VAR_COUNT);
}

int asmvars_value(const char key[], int *result)
{
    int hash_result =0;
    char err_msg[MAX_ERR_MSG_LEN];
    if (key == 0 || strlen(key) == 0)
    {
        fprintf(stderr, "INTERNAL_ERROR: asmvars_value: key is null");
        hash_result = -1;
    }
    else
    {
        hash_result = hash_value(key, result, asmvars, MAX_VAR_COUNT);
        if (hash_result == -1 && bkasm_stage > PARSE_STAGE)
        {
            sprintf(err_msg, "\nUndefined variable %s\n", key);
            throw_error(E_SYNTAXERROR, err_msg);
        }
    }
    return hash_result;
}

void asmvars_print()
{
    hash_print(asmvars);
}

File: bkasm/csharp/AsmVars.cs
using System;
using System.Collections.Generic;

namespace Bkasm
{
    public static class AsmVars
    {
        private static Dictionary<string, int> vars = new Dictionary<string, int>(Constants.MAX_VAR_COUNT);

        public static void Add(string key, int val)
        {
            if (string.IsNullOrEmpty(key))
            {
                Console.Error.WriteLine("INTERNAL_ERROR: asmvars_add: key is null");
                return;
            }

            if (Globals.bkasm_stage == BkasmStage.PARSE_STAGE)
            {
                if (vars.ContainsKey(key))
                {
                    ErrorHandler.ThrowError(eErrorType.E_LINKERERROR, "\nDuplicate label: " + key + "\n");
                    Globals.ExitNicely(eErrorType.E_LINKERERROR);
                    return;
                }
            }

            vars[key] = val;
        }

        public static bool Value(string key, out int result)
        {
            if (string.IsNullOrEmpty(key))
            {
                Console.Error.WriteLine("INTERNAL_ERROR: asmvars_value: key is null");
                result = 0;
                return false;
            }

            if (vars.TryGetValue(key, out result))
                return true;

            if (Globals.bkasm_stage > BkasmStage.EVAL_STAGE)
            {
                ErrorHandler.ThrowError(eErrorType.E_LINKERERROR, "\nUndefined variable " + key + "\n");
            }
            return false;
        }

        public static void Print()
        {
            var sorted = new List<KeyValuePair<string, int>>(vars);
            sorted.Sort((a, b) => string.Compare(a.Key, b.Key, StringComparison.Ordinal));
            foreach (var kv in sorted)
            {
                Console.WriteLine("{0} = {1}", kv.Key, kv.Value);
            }
        }
    }
}
